File: tools/topology-vscode/src/runner/counts.ts
import * as fs from "fs";
import * as path from "path";

function readCount(topologyPath: string, name: "nodes" | "edges"): number {
  const countPath = path.join(topologyPath, "counts", `${name}.json`);
  let raw: string;
  try {
    raw = fs.readFileSync(countPath, "utf8");
  } catch (err) {
    throw new Error(`cannot read ${countPath}: ${err instanceof Error ? err.message : String(err)}`);
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch (err) {
    throw new Error(`cannot parse ${countPath}: ${err instanceof Error ? err.message : String(err)}`);
  }
  if (typeof parsed !== "number" || !Number.isInteger(parsed) || parsed < 0) {
    throw new Error(`${countPath} must be a single non-negative integer, got: ${raw}`);
  }
  return parsed;
}

export function readCounts(topologyPath: string): { nodes: number; edges: number } {
  return { nodes: readCount(topologyPath, "nodes"), edges: readCount(topologyPath, "edges") };
}

File: Categories/Node/nodecrud/counts.go
package nodecrud

import (
	"path/filepath"
)

func nodesPath(root string) string { return filepath.Join(root, "counts", "nodes.bin") }

func edgesPath(root string) string { return filepath.Join(root, "counts", "edges.bin") }

func writeCounts(root string, nodes, edges int) error {
	if err := WriteAtomic(nodesPath(root), nodes); err != nil {
		return err
	}
	return WriteAtomic(edgesPath(root), edges)
}
